dango/testing/tests/dex_proptests.rs

```diff
@@ -0,0 +1,890 @@
+/// relative difference.
+fn assert_approx_eq(a: Uint128, b: Uint128, max_rel_diff: &str) -> Result<(), TestCaseError> {
+    // An absolute difference of up to one unit is acceptable, and unavoidable
+    // due to rounding errors. In this case, we consider the values effectively equal.
+    if absolute_difference(a, b) <= Uint128::ONE {
+        return Ok(());
+    }
+
+    // If the difference is greater than one unit, we ensure the relative difference
+    // isn't greater than a given threshold.
+    let rel_diff_num = relative_difference(a, b);
+    let rel_diff = Udec128::from_str(rel_diff_num.to_string().as_str()).unwrap();
+    prop_assert!(
+        rel_diff <= Udec128::from_str(max_rel_diff).unwrap(),
+        "assertion failed: values are not approximately equal\n  left: {}\n right: {}\n  max_rel_diff: {}\n  actual_rel_diff: {}",
+        a,
+        b,
+        max_rel_diff,
+        rel_diff
+    );
+
+    Ok(())
+}
+
+/// Helper function to register a fixed price for a denom
+fn register_fixed_price(
+    suite: &mut TestSuite<NaiveProposalPreparer>,
+    accounts: &mut TestAccounts,
+    contracts: &Contracts,
+    denom: Denom,
+    humanized_price: Udec128,
+    precision: u8,
+) {
+    // Register price source
+    suite
+        .execute(
+            &mut accounts.owner,
+            contracts.oracle,
+            &dango_types::oracle::ExecuteMsg::RegisterPriceSources(btree_map! {
+                denom => dango_types::oracle::PriceSource::Fixed {
+                    humanized_price,
+                    precision,
+                    // Use a very recent time to avoid the "price is too old" error.
+                    timestamp: MOCK_GENESIS_TIMESTAMP,
+                },
+            }),
+            Coins::default(),
+        )
+        .should_succeed();
+}
+
+/// Checks that the balances of the dex contract are equal to the balances of the open orders plus the balances of the passive liquidity.
+fn check_balances(
+    suite: &TestSuite<NaiveProposalPreparer>,
+    contracts: &Contracts,
+) -> Result<(), TestCaseError> {
+    // Check dex contract's balances.
+    let balances = suite.query_balances(&contracts.dex)?;
+    println!("dex contract balances: {balances:?}");
+
+    // Query the open orders.
+    let open_orders = suite.query_wasm_smart(contracts.dex, dex::QueryOrdersRequest {
+        start_after: None,
+        limit: None,
+    })?;
+    println!("open orders: {open_orders:?}");
+
+    let mut order_balances = Coins::new();
+    for (_, order) in open_orders {
+        let coin = if order.direction == Direction::Bid {
+            Coin::new(
+                order.quote_denom,
+                order.remaining.checked_mul_dec_ceil(order.price)?,
+            )?
+        } else {
+            Coin::new(order.base_denom, order.remaining)?
+        };
+        order_balances.insert(coin)?;
+    }
+    println!("order balances: {order_balances:?}");
```

Comment:
@apollo-sturdy for bids, the amount should be

```rust
order.remaining.checked_mul_dec_floor(order.price)?, // floor, not ceil!
```

This is because if you cancel the order now, the DEX contract uses floor to compute how many quote tokens to refund to you. See the `cancel_limit_order` function, in `dango/dex/src/execute/order_cancellation.rs`.

Similarly, in the margin account, in `compute_health`, we need to compute the value of limit orders. For those we should floor as well, for the same reason.

I think we can add a method `Order::remaining_tokens` as a source of truth for this operation. Let's do this in a separate PR.